libminifi/src/c2/C2Client.cpp

```diff
@@ -0,0 +1,318 @@
+      if (name.find(',') != std::string::npos) {
+        std::vector<std::string> sub_classes = utils::StringUtils::split(name, ",");
+        for (const std::string& subClassStr : classes) {
+          auto node = loadC2ResponseConfiguration(subClassStr, prev_node);
+          if (node != nullptr)
+            std::static_pointer_cast<state::response::ObjectNode>(prev_node)->add_node(node);
+        }
+      } else {
+        if (configuration_->get(classOption, class_definitions)) {
+          std::vector<std::string> classes = utils::StringUtils::split(class_definitions, ",");
+          for (const std::string& clazz : classes) {
+            // instantiate the object
+            auto ptr = core::ClassLoader::getDefaultClassLoader().instantiate(clazz, clazz);
+            if (nullptr == ptr) {
+              bool found_metric = [&] {
+                std::lock_guard<std::mutex> guard{metrics_mutex_};
+                auto metric = component_metrics_.find(clazz);
+                if (metric != component_metrics_.end()) {
+                  ptr = metric->second;
+                  return true;
+                }
+                return false;
+              }();
+              if (!found_metric) {
+                logger_->log_error("No metric defined for %s", clazz);
+                continue;
+              }
+            }
+
+            auto node = std::dynamic_pointer_cast<state::response::ResponseNode>(ptr);
+            std::static_pointer_cast<state::response::ObjectNode>(new_node)->add_node(node);
+          }
+          if (!new_node->isEmpty())
+            std::static_pointer_cast<state::response::ObjectNode>(prev_node)->add_node(new_node);
+
+        } else {
+          std::string optionName = option + "." + name;
+          auto sub_node = loadC2ResponseConfiguration(optionName, new_node);
+          std::static_pointer_cast<state::response::ObjectNode>(prev_node)->add_node(sub_node);
+        }
+      }
+    } catch (...) {
+      logger_->log_error("Could not create metrics class %s", metricsClass);
+    }
+  }
+  return prev_node;
+}
+
+std::shared_ptr<state::response::ResponseNode> C2Client::getMetricsNode(const std::string& metrics_class) const {
+  if (!metrics_class.empty()) {
+    std::lock_guard<std::mutex> lock(metrics_mutex_);
+    const auto citer = component_metrics_.find(metrics_class);
+    if (citer != component_metrics_.end()) {
+      return citer->second;
+    }
+  } else {
+    std::lock_guard<std::mutex> lock(metrics_mutex_);
+    const auto iter = root_response_nodes_.find("metrics");
+    if (iter != root_response_nodes_.end()) {
+      return iter->second;
+    }
+  }
+  return nullptr;
+}
+
+std::vector<std::shared_ptr<state::response::ResponseNode>> C2Client::getHeartbeatNodes(bool include_manifest) const {
+  std::string fullHb{"true"};
+  configuration_->get("nifi.c2.full.heartbeat", fullHb);
+  const bool include = include_manifest || fullHb == "true";
```

Comment:
The value of `fullHB` can only change here if a decryptor is set, is this what we want?

Reply:
I think at line 300 `fullHb`  will either contain the value of `nifi.c2.full.heartbeat` or `"true"` if that property is missing, how does the decryptor affect this?